### PR TITLE
[ibs-90] Make the Gitlab connection polymorphic

### PR DESCRIPTION
**Pull request by hezhongq**
## Description

- This PR introduces several changes aimed at enhancing the functionality of the version control system (VCS) integration within the project. The key changes include:
  1. Introduction of an abstract VCS class (`AbstractVCS`) along with concrete implementations for GitLab (`GitLabVCS`) and a mock VCS (`MockVCS`).
  2. Addition of a `user_id` column to the `User` model to facilitate user-related operations in the VCS.
  3. Mocking of certain functionalities in the `MockVCS` class to simulate GitLab API responses and facilitate testing.

## Changes Made

- This PR addresses the following tasks:
  1. Implement AbstractVCS and concrete VCS classes.
  2. Add user_id column to the User model.
  3. Implement VCS-related methods.
  4. Mock certain functionalities in the MockVCS class.

## Test Guidelines

1. Read through and review the newly added VCS files
2. Run basic tests such as `npm test registration`, `npm test login`, `npm test add_course`, `npm test get_course` etc to verify that the current login flow still works with the User model update.

### For PR author:
- [ ] I have reviewed my code.
- [ ] I have tested my changes locally.
- [ ] I have added appropriate comments to the code.
- [ ] I have ensured code style and formatting are consistent.
- [ ] I have included unit tests or integration tests where applicable.
- [ ] I have considered edge cases and handled them appropriately.
- [ ] I have celebrated my accomplishments 🎉

### For PR reviewer:
- [ ] I have reviewed the code changes.
- [ ] I have tested the changes locally.
- [ ] I have verified that comments are clear and helpful.
- [ ] I have ensured code style and formatting align with project standards.
- [ ] I have examined any included tests and their coverage.
- [ ] I have considered potential edge cases and provided feedback on them.

### CODE REVIEW

File: backend/lib/version_control/GitLabVCS.js [hunk 1]
```diff
@@ -0,0 +1,382 @@
+const AbstractVCS = require("./AbstractVCS");
+const db = require("../../setup/db");
+const axios = require("axios");
+
+class GitLabVCS extends AbstractVCS {
```
Comment (Kianoosh76) [inline backend/lib/version_control/GitLabVCS.js:5]: I think we should name it GitlabVCS instead of GitLab (lowercase l instead of L)

File: backend/lib/version_control/GitLabVCS.js [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+                console.log(err);
+            }
+            return { success: false, code: 'failed_remove_user' };
+        }
+
+        return { success: true };
+    }
+
+    /**
+     * gitlab_get_user_id
```
Comment (Kianoosh76) [inline backend/lib/version_control/GitLabVCS.js:296]: function name in docstring does not match the actual function name

File: backend/lib/version_control/AbstractVCS.js [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    static async remove_user_from_group(course_id, group_id, username) {
+        throw new Error("Method 'remove_user_from_group()' must be implemented.");
+    }
+
+    /**
+     * gitlab_get_user_id
+     * @param username
+     * @returns
+     */
+    static async get_vcs_user_id(username) {
```
Comment (Kianoosh76) [inline backend/lib/version_control/AbstractVCS.js:68]: maybe this function is better named get_user_id. It's already in VCS class so another mention of vcs in name is redundant

File: backend/lib/version_control/MockVCS.js [hunk 4]
```diff
@@ -0,0 +1,250 @@
+const AbstractVCS = require("./AbstractVCS");
+const db = require("../../setup/db");
+const axios = require("axios");
+const { User, Group, GroupUser, Course, Task } = require("../../models");
+
+class MockVCS extends AbstractVCS {
+
+    /**
+     * NEW gitlab_add_user_without_gitlab_group_id
```
Comment (Kianoosh76) [inline backend/lib/version_control/MockVCS.js:9]: this comment (and the rest in this file) should not mention gitlab

File: backend/utilities/helpers.js [hunk 5]
```diff
@@ -757,35 +759,35 @@ async function format_marks_all_tasks_csv(json, course_id, res, total) {
...
-            'message' in err['response']['data']
-        ) {
-            console.log(err['response']['data']['message']);
-        } else {
-            console.log(err);
-        }
-        return -1;
-    }
-}
+// async function gitlab_get_user_id(username) {
```
Comment (Kianoosh76) [inline backend/utilities/helpers.js:762]: Is there a reason this code is commented out instead of deleted?

File: backend/utilities/helpers.js [hunk 6]
```diff
@@ -757,35 +759,35 @@ async function format_marks_all_tasks_csv(json, course_id, res, total) {
...
+//         ) {
+//             console.log(err['response']['data']['message']);
+//         } else {
+//             console.log(err);
+//         }
+//         return -1;
+//     }
+// }
 
 async function gitlab_create_group_and_project_no_user(course_id, group_id, task) {
```
Comment (Kianoosh76) [inline backend/utilities/helpers.js:792]: should we remove these as well?

File: backend/lib/version_control/GitLabVCS.js [hunk 7]
```diff
@@ -0,0 +1,382 @@
...
+        await db.query(sql_add_gitlab_info, [
+            gitlab_subgroup_id,
+            gitlab_project_id,
+            gitlab_url,
+            group_id
+        ]);
+
+        return {
+            success: true,
+            gitlab_group_id: gitlab_subgroup_id,
```
Comment (Kianoosh76) [inline backend/lib/version_control/GitLabVCS.js:206]: Ideally the return values should not mention gitlab either, but let's not touch it for now

**Comment by hezhongq** [comment]
Verified the renaming of methods and comment removal, and the current login flow is not affected. Nice change, good to be merged.

**Review by DomiVesalius — APPROVED** [review]
